AlphaPass.sol

```diff
@@ -7,25 +7,63 @@ import "erc721a/contracts/ERC721A.sol";
 import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 
 contract AlphaNft is Ownable, ERC721A, ReentrancyGuard {
-    uint256 public immutable maxQty = 2000;
-    uint256 public immutable firstFree = 1000;
+    uint256 public immutable maxQty = 1024;
+    uint256 public immutable reserveQty = 300;
+    uint256 public tokensReserved;
+    // TODO: mintPrice to be decided
     uint256 public immutable mintPrice = 0.01 ether;
-    uint256 public immutable maxMintPerAddr = 10;
+    uint256 public immutable maxMintPerAddr = 1;
     string private _baseTokenURI;
 
-    // Flag whether address is minted or not
-    mapping(address => bool) private _isMinted;
+    //sale stages:
+    //stage 0: init(no minting, only reserve)
+    //stage 1: 200 whitelist mint
+    //stage 2: 200 whitelist mint, 100 public mint
+    //stage 3: 150 whitelist mint, 50 public mint
+    //stage 4: only reserve
+    uint8 public _stage = 0;
+    uint256 public immutable maxQtyStage1 = 200;
+    uint256 public immutable maxQtyStage2 = 300;
+    uint256 public immutable maxQtyStage3 = 200;
+    uint256 public _tokensMintedStage1 = 0;
+    uint256 public _tokensMintedStage2 = 0;
+    uint256 public _tokensMintedStage3 = 0;
+    bool public _isPublicMintOpen = false;
 
     constructor(string memory baseURI)
         ERC721A("AlphaPass", "AlphaPass")
     {
         _baseTokenURI = baseURI;
     }
 
+    function nextStage() external onlyOwner {
+        require(_stage <= 4, "Stage cannot be more than 4");
+        _stage++;
+        _isPublicMintOpen = false;
+    }
+
+    function setIsPublicMintOpen(bool isPublicMintOpen) external onlyOwner {
+        _isPublicMintOpen = isPublicMintOpen;
+    }
+
+    function reserve(address recipient, uint256 quantity) external onlyOwner {
```

Comment:
make it able to batch reserve (airdrop) and the func body.
`function reserve(address recipient, uint256 quantity) external onlyOwner`
=>
`function airdrop(address[] recipient, uint256[] quantity) external onlyOwner`